### PR TITLE
Added polls endpoints and events

### PR DESCRIPTION
**Pull request by martinmitrevski**
### 🎯 Goal

Setup support for building the networking part of the polls feature.

### 📝 Summary

Added requests/responses, endpoints and WS events for polls.

### 🛠 Implementation

This PR limits only to the payloads, the DB part and the controllers would be part of subsequent PRs.

### 🎨 Showcase

_Add relevant screenshots and/or videos/gifs to easily see what this PR changes, if applicable._

| Before | After |
| ------ | ----- |
|  img   |  img  |

### 🧪 Manual Testing Notes

Tested the requests manually, also added some unit tests. Nothing visual to test for now.

### ☑️ Contributor Checklist

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

ALT_TEXT

### CODE REVIEW

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift [hunk 1]
```diff
@@ -94,6 +96,8 @@ class MessagePayload: Decodable {
     var pinnedBy: UserPayload?
     var pinnedAt: Date?
     var pinExpires: Date?
+    
+    var poll: PollDTO?
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift:100]: DTOs should not be in Payloads. Why is this here?

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+struct PollPayload: Decodable {
+    let duration: String
+    let poll: PollResponseData
+
+    init(duration: String, poll: PollResponseData) {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:11]: Probably it would make more sense to switch the naming here. The higher level type is the Response, and the other is the actual payload. We have this same pattern in other places. I feel that `ResponseData` is a bit too generic to describe the structure of a type. Especially the `Data` post-fix, which we use sometimes if the type is binary.
```suggestion
struct PollPayloadResponse: Decodable {
    let duration: String
    let poll: PollPayload

    init(duration: String, poll: PollPayload) {
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case duration
+        case polls
+        case next
+        case prev
+    }
+}
+
+struct PollResponseData: Decodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:43]: As mentioned above, ideally, this would be `PollPayload` to keep consistency with the rest of the codebase. 
```suggestion
struct PollPayload: Decodable {
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    let createdById: String
+    let description: String
+    let enforceUniqueVote: Bool
+    let id: String
+    let name: String
+    let updatedAt: Date
+    let voteCount: Int
+    let votingVisibility: String
+    let options: [PollOptionResponseData?]
+    let ownVotes: [PollVoteResponseData?]
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:57]: We don't have this "ResponseData" postfix anywhere in the codebase, to make things consistent, all the payloads should have a `Payload` post-fix, ideally 
```suggestion
    let options: [PollOptionResponseData?]
    let ownVotes: [PollVoteResponseData?]
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        case voteCountsByOption = "vote_counts_by_option"
+        case isClosed = "is_closed"
+        case maxVotesAllowed = "max_votes_allowed"
+        case createdBy = "created_by"
+    }
+}
+
+struct PollOptionResponse: Decodable {
+    let duration: String
+    let pollOption: PollOptionResponseData
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:135]: From the comments above, this would be instead:
```suggestion
    let pollOption: PollOptionPayload
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        self.pollOption = pollOption
+    }
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case duration
+        case pollOption = "poll_option"
+    }
+}
+
+struct PollOptionResponseData: Decodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:148]: From the comments above, this would be instead:
```suggestion
struct PollOptionPayload: Decodable {
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case id
+        case text
+        case custom
+    }
+}
+
+struct PollVotePayload: Decodable {
+    var duration: String
+    var vote: PollVoteResponseData?
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:168]: From the comments above, this would be instead:
```suggestion
struct PollVoteResponse: Decodable {
    var duration: String
    var vote: PollVotePayload?
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case duration
+        case vote
+    }
+}
+
+struct PollVotesResponse: Decodable {
+    let duration: String
+    var votes: [PollVoteResponseData?]
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:183]: For lists, it would be nice to use `List` as a post-fix to be less confusing. Like we have `ThreadPayload` and `ThreadListPayload`. If we just use a `s` to distinguish both, it will cause confusion. `PollVotesPayload` and `PollVotePayload` are pretty much the same word, so it is easy to confuse both.

Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:183]: From the comments above, this would be instead:
```suggestion
struct PollVoteListResponse: Decodable {
    let duration: String
    var votes: [PollVotePayload?]
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case duration
+        case votes
+        case next
+        case prev
+    }
+}
+
+struct PollVoteResponseData: Decodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:207]:
```suggestion
struct PollVotePayload: Decodable {
```

File: Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+            path: .polls,
+            method: .post,
+            queryItems: nil,
+            requiresConnectionId: false,
+            requiresToken: true,
+            body: createPollRequest
+        )
+    }
+    
+    static func updatePoll(updatePollRequest: UpdatePollRequestBody) -> Endpoint<PollPayload> {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift:19]: This would be then:
`static func updatePoll(updatePollRequest: UpdatePollRequestBody) -> Endpoint<PollPayloadResponse> {`
or 
`static func updatePoll(updatePollRequest: UpdatePollRequestBody) -> Endpoint<PollResponse> {`

I'm fine with both `PollPayloadResponse` or `PollResponse`. What is important is that it has "Response" post-fix

File: Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+            path: .pollOptions(pollId: pollId),
+            method: .put,
+            queryItems: nil,
+            requiresConnectionId: false,
+            requiresToken: true,
+            body: updatePollOptionRequest
+        )
+    }
+    
+    static func deletePollOption(pollId: String, optionId: String) -> Endpoint<EmptyResponse> {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift:105]: You can see that here we have the "Response" post-fix

File: Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift [hunk 12]
```diff
@@ -0,0 +1,60 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+struct CastPollVoteRequestBody: Encodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift:7]: The request bodies look good and consistent with the rest of the codebase 👌

File: Sources/StreamChat/APIClient/Endpoints/Requests/CreatePollRequestBody.swift [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        case id
+        case isClosed = "is_closed"
+        case maxVotesAllowed = "max_votes_allowed"
+        case votingVisibility = "voting_visibility"
+        case options
+        case custom
+    }
+}
+
+struct PollOptionRequest: Encodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Requests/CreatePollRequestBody.swift:61]: nit: 
```suggestion
struct PollOptionRequestBody: Encodable {
```

File: Tests/StreamChatTests/APIClient/Endpoints/PollsEndpoints_Tests.swift [hunk 14]
```diff
@@ -0,0 +1,251 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+@testable import StreamChat
+@testable import StreamChatTestTools
+import XCTest
+
+final class PollsEndpoints_Tests: XCTestCase {
```
Comment (nuno-vieira) [inline Tests/StreamChatTests/APIClient/Endpoints/PollsEndpoints_Tests.swift:10]: It seems that we are missing PollPayload_Tests? Where we parse a real json response?

**Review by nuno-vieira — COMMENTED** [review]
Overall looks good 👍 I think some things should be improved:
- Ideally, the structures of the payloads, should have a `Payload` postfix, and the responses should have `Response` post-fix`.
- Removing the `ResponseData` post-fix, since we do not follow this pattern in the rest of the codebase
- Add Payload Parsing tests
- Use `List` to distinguish plural data (Although I guess this is optional, as long as the domain models use the `List` post-fix)

File: Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+    var latestAnswers: [PollVoteDTO?]
+    var options: [PollOptionDTO?]
+    var ownVotes: [PollVoteDTO?]
+    var custom: [String: RawJSON]?
+    var latestVotesByOption: [String: [PollVoteDTO?]]
+    var voteCountsByOption: [String: Int]
+    var isClosed: Bool?
+    var maxVotesAllowed: Int?
+    var votingVisibility: String?
+    var createdBy: UserPayload?
```
Comment (nuno-vieira) [inline Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift:98]: This should be Payload. DTO's (in our codebase) are used for CoreData models
```suggestion
struct PollPayload: Decodable {
    var allowAnswers: Bool
    var allowUserSuggestedOptions: Bool
    var answersCount: Int
    var createdAt: Date
    var createdById: String
    var description: String
    var enforceUniqueVote: Bool
    var id: String
    var name: String
    var updatedAt: Date
    var voteCount: Int
    var latestAnswers: [PollVoteDTO?]
    var options: [PollOptionDTO?]
    var ownVotes: [PollVoteDTO?]
    var custom: [String: RawJSON]?
    var latestVotesByOption: [String: [PollVoteDTO?]]
    var voteCountsByOption: [String: Int]
    var isClosed: Bool?
    var maxVotesAllowed: Int?
    var votingVisibility: String?
    var createdBy: UserPayload?
```

File: Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift (hunk 1 above)
Comment (martinmitrevski) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/MessagePayloads.swift:100]: yeah, I was mostly following the REST API naming - let me fix those.

File: Tests/StreamChatTests/APIClient/Endpoints/PollsEndpoints_Tests.swift (hunk 14 above)
Comment (martinmitrevski) [reply Tests/StreamChatTests/APIClient/Endpoints/PollsEndpoints_Tests.swift:10]: added

File: Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift (hunk 11 above)
Comment (martinmitrevski) [reply Sources/StreamChat/APIClient/Endpoints/PollsEndpoints.swift:105]: fixed all these, I was following the backend conventions the first time

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift (hunk 8 above)
Comment (martinmitrevski) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift]: fixed

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+        self.poll = poll
+    }
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case duration
+        case poll
+    }
+}
+
+struct QueryPollsPayloadResponse: Decodable {
```
Comment (nuno-vieira) [inline Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift:22]: nit: Shouldn't this be PollsList? 🤔 
```suggestion
struct PollsListPayloadResponse: Decodable {
```

File: Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+    init(payload: EventPayload, vote: PollVotePayload? = nil) {
+        self.payload = payload
+        self.vote = vote
+    }
+}
+
+struct PollPayload: Decodable {
+    var allowAnswers: Bool
+    var allowUserSuggestedOptions: Bool
```
Comment (nuno-vieira) [inline Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift:79]: It is probably better to move this to the Payloads file

File: TestTools/StreamChatTestTools/Fixtures/JSONs/Poll.json [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        "enforce_unique_vote": false,
+        "max_votes_allowed": null,
+        "allow_user_suggested_options": false,
+        "allow_answers": false,
+        "vote_count": 0,
+        "options": [],
+        "answers_count": 0,
+        "vote_counts_by_option": null,
+        "latest_votes_by_option": {},
+        "own_votes": [],
```
Comment (nuno-vieira) [inline TestTools/StreamChatTestTools/Fixtures/JSONs/Poll.json:16]: We are not covering the Votes and Options parsing it seems?

File: hooks/pre-commit.sh [hunk 19]
```diff
@@ -1,10 +1,6 @@
 #!/bin/bash
 set -eo pipefail
 
-if [ "$GITHUB_ACTIONS" != true ]; then
-  vale docusaurus
-fi
```
Comment (nuno-vieira) [inline hooks/pre-commit.sh:6]: Why are we doing this? 🤔

File: Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    ) {
+        self.answerText = answerText
+        self.optionId = optionId
+        self.option = option
+    }
+    
+    enum CodingKeys: String, CodingKey, CaseIterable {
+        case answerText = "answer_text"
+        case optionId = "option_id"
+        case option = "Option"
```
Comment (laevandus) [inline Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift:40]: I guess someone make a mistake on the backend side with capitalisation

File: Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift (hunk 16 above)
Comment (martinmitrevski) [reply Sources/StreamChat/APIClient/Endpoints/Payloads/PollsPayloads.swift]: done

File: Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift (hunk 17 above)
Comment (martinmitrevski) [reply Sources/StreamChat/WebSocketClient/Events/PollsEvents.swift]: done

File: TestTools/StreamChatTestTools/Fixtures/JSONs/Poll.json (hunk 18 above)
Comment (martinmitrevski) [reply TestTools/StreamChatTestTools/Fixtures/JSONs/Poll.json]: added

File: hooks/pre-commit.sh (hunk 19 above)
Comment (martinmitrevski) [reply hooks/pre-commit.sh:6]: this breaks the workflow (can't commit) if you don't have a specific vale version, which is no longer supported. Leaving this for @testableapple to solve properly.

File: Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift (hunk 20 above)
Comment (martinmitrevski) [reply Sources/StreamChat/APIClient/Endpoints/Requests/CastPollVoteRequestBody.swift:40]: It's correct with capital O: https://getstream.io/chat/docs/rest/#polls-castpollvote. 😁

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅
